src/System.CommandLine/Parsing/ParseResultVisitor.cs

```diff
@@ -81,7 +81,7 @@ protected override void VisitCommandNode(CommandNode commandNode)
             var commandResult = _innermostCommandResult;
 
             var argumentResult =
-                commandResult.Children
+                commandResult!.Children
```

Comment:
consider assertions for program invariants, as opposed to checks in the current method body which the compiler didn't or couldn't understand.

Reply:
@jcouv Just to make sure I understand. You are suggesting that it is better to leverage Debug.Assert here rather than the null forgiveness operator? Similar to your answer [here](https://stackoverflow.com/a/59701577/3755169)

Reply:
Yes, that's what I'm suggesting.

Reply:
Perhaps I am doing something wrong here, but when I place `Debug.Assert(commandResult != null)` directly above this line and remove the null forgiveness operator, I still get the warning. Is this because the project is targeting netstandard2.0 (meaning the annotations that were added on Debug.Assert are not there)?

Reply:
@jcouv I typically avoid `Debug.Assert` because it will cause the entire test run to fail in CI with no actionable feedback. Thoughts?

Reply:
@jonsequitur Ouch. That would make assertions painful indeed :( 
We use a lot of assertions in Roslyn and I never noticed it failing entire test suits. Is this behavior new with Core vs. desktop? Is it the same with a NullRefException?

The nullability feature relies a significantly on `Debug.Assert` to give the compiler hints, so it might be worthwhile to make it behave in tests (ie. fail a single test rather than the entire suite).

Reply:
I've been thinking about this problem. Maybe an option is to add your own assertion method with the `[DoesNotReturnIf(false)]` annotation, but don't have it actually assert anything until the CI issue could be resolved (exceptions not being reported on individual test that failed). 
That one helper method could be marked with follow-up issue and eventually be made into a bonafide assertion.

Reply:
On .NET Core it crashes the process and leaves an empty `.trx` file (if you're using `--logger trx`). If someone happens to be relying on `.trx` file output instead of the test runner result code, this can lead to an incorrect pass on the test run. On .NET Framework, it pops a modal dialog and hangs the build.

These behaviors can be customized by removing the `DefaultTraceListener` and, for example, replacing it with a `TraceListener` that throws an exception, but this is kind of a non-obvious ritual.